Emulator snapshot managment

The CLI should support the new emulator snapshot management feature developed by @bluesign

Emulator exposes a new emulator admin API on `/emulator` which exposes new functionality such as snapshot management. You can save a new snapshot by calling `/emulator/snapshot/{name}` and retrieve a snapshot by name if already exists on that same endpoint. Besides that it enables creating new blocks on demand by calling `/emulator/newBlock`. 

This API should be supported and exposed by the CLI as `emulator` subcommand. My thinking here is `flow emulator blocks commit` and `flow emulator snapshot {name}`.

Currently doesn't seem like we have the API implemented for saving snapshot for emulator

````
Raymonds-MBP-3:flow raymondsukanto$ ./flow snapshot save /tmp/snapshot.json
networkey we are in gateway⚠️ warning: using insecure client connection to download snapshot, you should use a secure network configuration...
❌ Command Error: failed to get latest finalized protocol snapshot from gateway: client: rpc error: code = Unavailable desc = error reading from server: EOF
````

Terminal running emulator exited due to panic
````
Raymonds-MBP-3:flow raymondsukanto$ ./flow emulator
INFO[0000] ⚙️   Using service account 0xf8d6e0586b0a20c7  serviceAddress=f8d6e0586b0a20c7 serviceHashAlgo=SHA3_256 servicePrivKey=e2f55d130f6aada644c315436a8faee35c2e7d42f8c9fd4e94761b3e00119cc9 servicePubKey=3d25391c2e8a6def0c1786de192bc3022e086c8db85059beedf24f5177af3a8d4ecab8eb3862cb411ff56e243c354e171da575d20f253d50ff52939087ea6cca serviceSigAlgo=ECDSA_P256
INFO[0000] 📜  Flow contract                              FlowServiceAccount=0xf8d6e0586b0a20c7
INFO[0000] 📜  Flow contract                              FlowToken=0x0ae53cb6e3f42a79
INFO[0000] 📜  Flow contract                              FungibleToken=0xee82856bf20e2aa6
INFO[0000] 📜  Flow contract                              FlowFees=0xe5a8b7f23e8b548f
INFO[0000] 📜  Flow contract                              FlowStorageFees=0xf8d6e0586b0a20c7
INFO[0000] 🌱  Starting gRPC server on port 3569          port=3569
INFO[0000] 🌱  Starting REST API on port 8888             port=8888
INFO[0000] 🌱  Starting admin server on port 8080         port=8080
panic: implement me

goroutine 55 [running]:
github.com/onflow/flow-emulator/server/backend.(*Backend).GetLatestProtocolStateSnapshot(...)
        /Users/raymondsukanto/go/pkg/mod/github.com/onflow/flow-emulator@v0.33.2/server/backend/backend.go:585
github.com/onflow/flow-emulator/server/backend.(*Adapter).GetLatestProtocolStateSnapshot(0x140013199d8?, {0x1013109ac?, 0x1020ca000?})
````

--- a/internal/emulator/snapshot.go
+++ b/internal/emulator/snapshot.go
@@ -22,13 +22,15 @@ import (
 	"bytes"
 	"encoding/json"
 	"fmt"
+	"io"
+	"net/http"
+
 	"github.com/onflow/flow-cli/internal/command"
 	"github.com/onflow/flow-cli/pkg/flowkit"
 	"github.com/onflow/flow-cli/pkg/flowkit/services"
 	"github.com/onflow/flow-cli/pkg/flowkit/util"
+
 	"github.com/spf13/cobra"
-	"io"
-	"net/http"
 )
 
 type SnapshotFlag struct {
@@ -51,7 +53,7 @@ type SnapShotResult struct {
 	Name    string `json:"context"`
 	BlockID string `json:"blockId"`
 	Height  uint64 `json:"height"`
-	Result  string `json:"height,omitempty"`
+	Result  string `json:"result,omitempty"`
 }
 
 func (r *SnapShotResult) JSON() interface{} {

--- a/internal/emulator/snapshot.go
+++ b/internal/emulator/snapshot.go
@@ -22,7 +22,6 @@ import (
 	"bytes"
 	"encoding/json"
 	"fmt"
-	"golang.org/x/exp/slices"
 	"io"
 	"net/http"
 	"strings"
@@ -33,6 +32,7 @@ import (
 	"github.com/onflow/flow-cli/pkg/flowkit/util"
 
 	"github.com/spf13/cobra"
+	"golang.org/x/exp/slices"
 )
 
 type SnapshotFlag struct {

--- a/internal/emulator/snapshot.go
+++ b/internal/emulator/snapshot.go
@@ -22,8 +22,10 @@ import (
 	"bytes"
 	"encoding/json"
 	"fmt"
+	"golang.org/x/exp/slices"
 	"io"
 	"net/http"
+	"strings"
 
 	"github.com/onflow/flow-cli/internal/command"
 	"github.com/onflow/flow-cli/pkg/flowkit"
@@ -40,13 +42,36 @@ var snapshotFlag = SnapshotFlag{}
 
 var SnapshotCmd = &command.Command{
 	Cmd: &cobra.Command{
-		Use:     "snapshot <snapshotName>",
-		Short:   "Save/Load a emulator snapshot",
-		Example: "flow emulator snapshot testSnapshot",
-		Args:    cobra.ExactArgs(1),
+		Use:     "snapshot <create|load|list> [snapshotName]",
+		Short:   "Create/Load/List emulator snapshots",
+		Example: "flow emulator snapshot create testSnapshot",
+		Args:    cobra.RangeArgs(1, 2),
 	},
 	Flags: &snapshotFlag,
 	Run:   snapshot,
+}
+
+type SnapshotList struct {
+	Snapshots []string
+}
+
+func (s *SnapshotList) JSON() interface{} {
+	return s.Snapshots
+}
+
+func (s *SnapshotList) String() string {
+	var b bytes.Buffer
+	writer := util.CreateTabWriter(&b)
+	_, _ = fmt.Fprintf(writer, "Snapshots:\n")
+	for _, snapshotName := range s.Snapshots {
+		_, _ = fmt.Fprintf(writer, "\t%s\n", snapshotName)
+	}
+	_ = writer.Flush()
+	return b.String()
+}
+
+func (s *SnapshotList) Oneliner() string {
+	return strings.Join(s.Snapshots, ",")
 }
 
 type SnapShotResult struct {
@@ -86,29 +111,78 @@ func (r *SnapShotResult) Oneliner() string {
 	return fmt.Sprintf("%s : %s (%d) %s", r.Name, r.BlockID, r.Height, r.Result)
 }
 
+const SnapshotEndpoint = "http://localhost:8080/emulator/snapshots"
+
+func makeRequest(r *http.Request, v any) error {
+	resp, err := http.DefaultClient.Do(r)
+	if err != nil {
+		return fmt.Errorf("emulator snapshot request error: %w", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("emulator snapshot request error: status_code=%d", resp.StatusCode)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	err = json.Unmarshal(body, v)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func requestListSnapshots() (*http.Request, error) {
+	request, err := http.NewRequest("GET", SnapshotEndpoint, nil)
+	if err != nil {
+		return nil, err
+	}
+	return request, nil
+}
+
+func requestCreateSnapshot(name string) (*http.Request, error) {
+	requestBody := bytes.NewBufferString(fmt.Sprintf("name=%s", name))
+	request, err := http.NewRequest("POST", SnapshotEndpoint, requestBody)
+	request.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	if err != nil {
+		return nil, err
+	}
+	return request, nil
+}
+
+func requestLoadSnapshot(name string) (*http.Request, error) {
+	request, err := http.NewRequest("PUT", fmt.Sprintf("%s/%s", SnapshotEndpoint, name), nil)
+	if err != nil {
+		return nil, err
+	}
+	return request, nil
+}
+
 func listSnapshot() (result []string, err error) {
-
-	snapshots, err := http.Get("http://localhost:8080/emulator/snapshots")
+	req, err := requestListSnapshots()
 	if err != nil {
 		return []string{}, err
 	}
 
-	if snapshots.StatusCode != http.StatusOK {
-		return []string{}, fmt.Errorf("unable to list snapshots on the emulator")
-	}
-
-	body, err := io.ReadAll(snapshots.Body)
-	if err != nil {
-		return []string{}, err
-	}
-
-	err = json.Unmarshal(body, &result)
+	err = makeRequest(req, &result)
 	if err != nil {
 		return []string{}, err
 	}
 
 	return result, nil
 }
+
+type SnapshotCommand string
+
+const (
+	SnapshotCommandList   SnapshotCommand = "list"
+	SnapshotCommandCreate SnapshotCommand = "create"
+	SnapshotCommandLoad   SnapshotCommand = "load"
+)
 
 func snapshot(
 	args []string,
@@ -117,80 +191,63 @@ func snapshot(
 	_ *services.Services,
 ) (command.Result, error) {
 
-	name := args[0]
+	subCommand := args[0]
 
 	snapshots, err := listSnapshot()
 	if err != nil {
 		return nil, err
 	}
 
-	exists := false
-	for _, snapshotName := range snapshots {
-		if name == snapshotName {
-			exists = true
-			break
+	switch SnapshotCommand(subCommand) {
+	case SnapshotCommandList:
+		return &SnapshotList{Snapshots: snapshots}, nil
+
+	case SnapshotCommandCreate:
+		if len(args) < 2 {
+			return nil, fmt.Errorf("snapshot create command requires name argument")
 		}
-	}
-
-	if exists {
-		//load snapshot
-		requestBody := bytes.NewBufferString("")
-		request, err := http.NewRequest("PUT",
-			fmt.Sprintf("http://localhost:8080/emulator/snapshots/%s", name),
-			requestBody)
-
-		if err != nil {
-			return nil, err
-		}
-		resp, err := http.DefaultClient.Do(request)
-
-		if err != nil {
-			return nil, err
-		}
-
-		body, err := io.ReadAll(resp.Body)
-		if err != nil {
-			return nil, err
-		}
-
-		if resp.StatusCode != http.StatusOK {
-			return nil, fmt.Errorf("unable to create snapshot on the emulator")
+		name := args[1]
+		exists := slices.Contains(snapshots, name)
+		if exists {
+			return nil, fmt.Errorf("snapshot '%s' already exists", name)
 		}
 
 		var result SnapShotResult
-		err = json.Unmarshal(body, &result)
+		req, err := requestCreateSnapshot(name)
 		if err != nil {
 			return nil, err
 		}
-		result.Result = "Snapshot loaded"
-		return &result, nil
-	} else {
-		//save snapshot
-		requestBody := bytes.NewBufferString(fmt.Sprintf("name=%s", name))
-		resp, err := http.Post("http://localhost:8080/emulator/snapshots",
-			"application/x-www-form-urlencoded",
-			requestBody)
-
-		if err != nil {
-			return nil, err
-		}
-
-		body, err := io.ReadAll(resp.Body)
-		if err != nil {
-			return nil, err
-		}
-
-		if resp.StatusCode != http.StatusOK {
-			return nil, fmt.Errorf("unable to create snapshot on the emulator")
-		}
-
-		var result SnapShotResult
-		err = json.Unmarshal(body, &result)
+		err = makeRequest(req, &result)
 		if err != nil {
 			return nil, err
 		}
 		result.Result = "Snapshot created"
 		return &result, nil
+
+	case SnapshotCommandLoad:
+		if len(args) < 2 {
+			return nil, fmt.Errorf("snapshot load command requires name argument")
+		}
+		name := args[1]
+		exists := slices.Contains(snapshots, name)
+		if !exists {
+			return nil, fmt.Errorf("snapshot '%s' does not exist", name)
+		}
+
+		var result SnapShotResult
+		req, err := requestLoadSnapshot(name)
+		if err != nil {
+			return nil, err
+		}
+		err = makeRequest(req, &result)
+		if err != nil {
+			return nil, err
+		}
+		result.Result = "Snapshot loaded"
+		return &result, nil
+
+	default:
+		return nil, fmt.Errorf("invalid snapshot command: valid commands are: 'list', 'create', 'load'")
 	}
 
 }

--- a/internal/emulator/snapshot.go
+++ b/internal/emulator/snapshot.go
@@ -1,0 +1,194 @@
+/*
+ * Flow CLI
+ *
+ * Copyright 2019 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package emulator
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/onflow/flow-cli/internal/command"
+	"github.com/onflow/flow-cli/pkg/flowkit"
+	"github.com/onflow/flow-cli/pkg/flowkit/services"
+	"github.com/onflow/flow-cli/pkg/flowkit/util"
+	"github.com/spf13/cobra"
+	"io"
+	"net/http"
+)
+
+type SnapshotFlag struct {
+}
+
+var snapshotFlag = SnapshotFlag{}
+
+var SnapshotCmd = &command.Command{
+	Cmd: &cobra.Command{
+		Use:     "snapshot <snapshotName>",
+		Short:   "Save/Load a emulator snapshot",
+		Example: "flow emulator snapshot testSnapshot",
+		Args:    cobra.ExactArgs(1),
+	},
+	Flags: &snapshotFlag,
+	Run:   snapshot,
+}
+
+type SnapShotResult struct {
+	Name    string `json:"context"`
+	BlockID string `json:"blockId"`
+	Height  uint64 `json:"height"`
+	Result  string `json:"height,omitempty"`
+}
+
+func (r *SnapShotResult) JSON() interface{} {
+	result := make(map[string]interface{})
+	result["name"] = r.Name
+	result["blockID"] = r.BlockID
+	result["height"] = r.Height
+	if r.Result != "" {
+		result["result"] = r.Result
+	}
+
+	return result
+}
+
+func (r *SnapShotResult) String() string {
+	var b bytes.Buffer
+	writer := util.CreateTabWriter(&b)
+	if r.Result != "" {
+		_, _ = fmt.Fprintf(writer, "%s\n", r.Result)
+	}
+	_, _ = fmt.Fprintf(writer, "Name\t%s\n", r.Name)
+	_, _ = fmt.Fprintf(writer, "Block ID\t%s\n", r.BlockID)
+	_, _ = fmt.Fprintf(writer, "Height\t%d", r.Height)
+
+	_ = writer.Flush()
+	return b.String()
+}
+
+func (r *SnapShotResult) Oneliner() string {
+	return fmt.Sprintf("%s : %s (%d) %s", r.Name, r.BlockID, r.Height, r.Result)
+}
+
+func listSnapshot() (result []string, err error) {
+
+	snapshots, err := http.Get("http://localhost:8080/emulator/snapshots")
+	if err != nil {
+		return []string{}, err
+	}
+
+	if snapshots.StatusCode != http.StatusOK {
+		return []string{}, fmt.Errorf("unable to list snapshots on the emulator")
+	}
+
+	body, err := io.ReadAll(snapshots.Body)
+	if err != nil {
+		return []string{}, err
+	}
+
+	err = json.Unmarshal(body, &result)
+	if err != nil {
+		return []string{}, err
+	}
+
+	return result, nil
+}
+
+func snapshot(
+	args []string,
+	_ flowkit.ReaderWriter,
+	_ command.GlobalFlags,
+	_ *services.Services,
+) (command.Result, error) {
+
+	name := args[0]
+
+	snapshots, err := listSnapshot()
+	if err != nil {
+		return nil, err
+	}
+
+	exists := false
+	for _, snapshotName := range snapshots {
+		if name == snapshotName {
+			exists = true
+			break
+		}
+	}
+
+	if exists {
+		//load snapshot
+		requestBody := bytes.NewBufferString("")
+		request, err := http.NewRequest("PUT",
+			fmt.Sprintf("http://localhost:8080/emulator/snapshots/%s", name),
+			requestBody)
+
+		if err != nil {
+			return nil, err
+		}
+		resp, err := http.DefaultClient.Do(request)
+
+		if err != nil {
+			return nil, err
+		}
+
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		if resp.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("unable to create snapshot on the emulator")
+		}
+
+		var result SnapShotResult
+		err = json.Unmarshal(body, &result)
+		if err != nil {
+			return nil, err
+		}
+		result.Result = "Snapshot loaded"
+		return &result, nil
+	} else {
+		//save snapshot
+		requestBody := bytes.NewBufferString(fmt.Sprintf("name=%s", name))
+		resp, err := http.Post("http://localhost:8080/emulator/snapshots",
+			"application/x-www-form-urlencoded",
+			requestBody)
+
+		if err != nil {
+			return nil, err
+		}
+
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		if resp.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("unable to create snapshot on the emulator")
+		}
+
+		var result SnapShotResult
+		err = json.Unmarshal(body, &result)
+		if err != nil {
+			return nil, err
+		}
+		result.Result = "Snapshot created"
+		return &result, nil
+	}
+
+}

--- a/internal/emulator/start.go
+++ b/internal/emulator/start.go
@@ -110,6 +110,7 @@ func init() {
 	Cmd.Use = "emulator"
 	Cmd.Short = "Run Flow network for development"
 	Cmd.GroupID = "tools"
+	SnapshotCmd.AddToParent(Cmd)
 }
 
 func Exitf(code int, msg string, args ...interface{}) {

